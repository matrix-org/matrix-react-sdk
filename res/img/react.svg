<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
