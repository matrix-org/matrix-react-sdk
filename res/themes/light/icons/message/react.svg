<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>