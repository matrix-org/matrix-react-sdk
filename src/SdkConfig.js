/*
Copyright 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

const DEFAULTS = {
    // URL to a page we show in an iframe to configure integrations
    integrations_ui_url: "https://scalar.vector.im/",
    // Base URL to the REST interface of the integrations server
    integrations_rest_url: "https://scalar.vector.im/api",
    // Where to send bug reports. If not specified, bugs cannot be sent.
    bug_report_endpoint_url: null,
};

class SdkConfig {
    static get() {
        return global.mxReactSdkConfig || {};
    }

    static put(cfg) {
        const defaultKeys = Object.keys(DEFAULTS);
        for (let i = 0; i < defaultKeys.length; ++i) {
            if (cfg[defaultKeys[i]] === undefined) {
                cfg[defaultKeys[i]] = DEFAULTS[defaultKeys[i]];
            }
        }
        global.mxReactSdkConfig = cfg;
    }

    static unset() {
        global.mxReactSdkConfig = undefined;
    }
}

module.exports = SdkConfig;
