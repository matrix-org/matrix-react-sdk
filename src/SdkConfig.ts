/*
Copyright 2016 OpenMarket Ltd
Copyright 2019 - 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { Optional } from "matrix-events-sdk";

import { SnakedObject } from "./utils/SnakedObject";
import { IConfigOptions, ISsoRedirectOptions } from "./IConfigOptions";
import { KeysWithObjectShape } from "./@types/common";

// see element-web config.md for docs, or the IConfigOptions interface for dev docs
export const DEFAULTS: IConfigOptions = {
    brand: "Element",
    integrations_ui_url: "https://scalar.vector.im/",
    integrations_rest_url: "https://scalar.vector.im/api",
    bug_report_endpoint_url: null,
    jitsi: {
        preferred_domain: "meet.element.io",
    },

    // @ts-ignore - we deliberately use the camelCase version here so we trigger
    // the fallback behaviour. If we used the snake_case version then we'd break
    // everyone's config which has the camelCase property because our default would
    // be preferred over their config.
    desktopBuilds: {
        available: true,
        logo: require("../res/img/element-desktop-logo.svg").default,
        url: "https://element.io/get-started",
    },
    spaces_learn_more_url: "https://element.io/blog/spaces-blast-out-of-beta/",
};

export default class SdkConfig {
    private static instance: IConfigOptions;
    private static fallback: SnakedObject<IConfigOptions>;

    private static setInstance(i: IConfigOptions) {
        SdkConfig.instance = i;
        SdkConfig.fallback = new SnakedObject(i);

        // For debugging purposes
        window.mxReactSdkConfig = i;
    }

    public static get(): IConfigOptions;
    public static get<K extends keyof IConfigOptions>(key: K, altCaseName?: string): IConfigOptions[K];
    public static get<K extends keyof IConfigOptions = never>(
        key?: K, altCaseName?: string,
    ): IConfigOptions | IConfigOptions[K] {
        if (key === undefined) {
            // safe to cast as a fallback - we want to break the runtime contract in this case
            return SdkConfig.instance || <IConfigOptions>{};
        }
        return SdkConfig.fallback.get(key, altCaseName);
    }

    public static getObject<K extends KeysWithObjectShape<IConfigOptions>>(
        key: K, altCaseName?: string,
    ): Optional<SnakedObject<IConfigOptions[K]>> {
        const val = SdkConfig.get(key, altCaseName);
        if (val !== null && val !== undefined) {
            return new SnakedObject(val);
        }

        // return the same type for sensitive callers (some want `undefined` specifically)
        return val === undefined ? undefined : null;
    }

    public static put(cfg: IConfigOptions) {
        const defaultKeys = Object.keys(DEFAULTS);
        for (let i = 0; i < defaultKeys.length; ++i) {
            if (cfg[defaultKeys[i]] === undefined) {
                cfg[defaultKeys[i]] = DEFAULTS[defaultKeys[i]];
            }
        }
        SdkConfig.setInstance(cfg);
    }

    /**
     * Resets the config to be completely empty.
     */
    public static unset() {
        SdkConfig.setInstance(<IConfigOptions>{}); // safe to cast - defaults will be applied
    }

    public static add(cfg: Partial<IConfigOptions>) {
        const liveConfig = SdkConfig.get();
        const newConfig = Object.assign({}, liveConfig, cfg);
        SdkConfig.put(newConfig);
    }
}

export function parseSsoRedirectOptions(config: IConfigOptions): ISsoRedirectOptions {
    // Ignore deprecated options if the config is using new ones
    if (config.sso_redirect_options) return config.sso_redirect_options;

    // We can cheat here because the default is false anyways
    if (config.sso_immediate_redirect) return { immediate: true };

    // Default: do nothing
    return {};
}
