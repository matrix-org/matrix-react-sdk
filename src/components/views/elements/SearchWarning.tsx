/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { logger } from "matrix-js-sdk/src/logger";

import EventIndexPeg from "../../../indexing/EventIndexPeg";
import { _t } from "../../../languageHandler";
import SdkConfig from "../../../SdkConfig";
import dis from "../../../dispatcher/dispatcher";
import { Action } from "../../../dispatcher/actions";
import { UserTab } from "../dialogs/UserTab";
import AccessibleButton from "./AccessibleButton";

export enum WarningKind {
    Files,
    Search,
}

interface IProps {
    isRoomEncrypted: boolean;
    kind: WarningKind;
}

export default function SearchWarning({ isRoomEncrypted, kind }: IProps) {
    if (!isRoomEncrypted) return null;
    if (EventIndexPeg.get()) return null;

    if (EventIndexPeg.error) {
        return (
            <div className="mx_SearchWarning">
                { _t("Message search initialisation failed, check <a>your settings</a> for more information", {}, {
                    a: sub => (
                        <AccessibleButton
                            kind="link_inline"
                            onClick={(evt) => {
                                evt.preventDefault();
                                dis.dispatch({
                                    action: Action.ViewUserSettings,
                                    initialTabId: UserTab.Security,
                                });
                            }}
                        >
                            { sub }
                        </AccessibleButton>),
                }) }
            </div>
        );
    }

    const brand = SdkConfig.get("brand");
    const desktopBuilds = SdkConfig.getObject("desktop_builds");

    let text = null;
    let logo = null;
    if (desktopBuilds.get("available")) {
        logo = <img src={desktopBuilds.get("logo")} />;
        const buildUrl = desktopBuilds.get("url");
        switch (kind) {
            case WarningKind.Files:
                text = _t("Use the <a>Desktop app</a> to see all encrypted files", {}, {
                    a: sub => (<a href={buildUrl} target="_blank" rel="noreferrer noopener">{ sub }</a>),
                });
                break;
            case WarningKind.Search:
                text = _t("Use the <a>Desktop app</a> to search encrypted messages", {}, {
                    a: sub => (<a href={buildUrl} target="_blank" rel="noreferrer noopener">{ sub }</a>),
                });
                break;
        }
    } else {
        switch (kind) {
            case WarningKind.Files:
                text = _t("This version of %(brand)s does not support viewing some encrypted files", { brand });
                break;
            case WarningKind.Search:
                text = _t("This version of %(brand)s does not support searching encrypted messages", { brand });
                break;
        }
    }

    // for safety
    if (!text) {
        logger.warn("Unknown desktop builds warning kind: ", kind);
        return null;
    }

    return (
        <div className="mx_SearchWarning">
            { logo }
            <span>{ text }</span>
        </div>
    );
}
