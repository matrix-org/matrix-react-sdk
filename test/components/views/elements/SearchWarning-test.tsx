/*
Copyright 2024 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { render } from "@testing-library/react";
import React from "react";

import SdkConfig from "../../../../src/SdkConfig";
import SearchWarning, { WarningKind } from "../../../../src/components/views/elements/SearchWarning";

describe("<SearchWarning />", () => {
    describe("with desktop builds available", () => {
        beforeEach(() => {
            SdkConfig.put({
                brand: "Element",
                desktop_builds: {
                    available: true,
                    logo: "https://logo",
                    url: "https://url",
                },
            });
        });

        it("renders with a logo by default", () => {
            const { asFragment, getByRole } = render(
                <SearchWarning isRoomEncrypted={true} kind={WarningKind.Search} />,
            );
            expect(getByRole("img")).toHaveAttribute("src", "https://logo");
            expect(asFragment()).toMatchSnapshot();
        });

        it("renders without a logo when showLogo=false", () => {
            const { asFragment, queryByRole } = render(
                <SearchWarning isRoomEncrypted={true} kind={WarningKind.Search} showLogo={false} />,
            );

            expect(queryByRole("img")).not.toBeInTheDocument();
            expect(asFragment()).toMatchSnapshot();
        });
    });
});
