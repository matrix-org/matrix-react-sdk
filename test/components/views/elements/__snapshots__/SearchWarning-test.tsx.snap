// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchWarning /> with desktop builds available renders with a logo by default 1`] = `
<DocumentFragment>
  <div
    class="mx_SearchWarning"
  >
    <img
      alt=""
      src="https://logo"
      width="32px"
    />
    <span>
      <span>
        <a
          href="https://url"
          rel="noreferrer noopener"
          target="_blank"
        >
          Open the Desktop app
        </a>
         to search rooms with encrypted messages
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`<SearchWarning /> with desktop builds available renders without a logo when showLogo=false 1`] = `
<DocumentFragment>
  <div
    class="mx_SearchWarning"
  >
    <span>
      <span>
        <a
          href="https://url"
          rel="noreferrer noopener"
          target="_blank"
        >
          Open the Desktop app
        </a>
         to search rooms with encrypted messages
      </span>
    </span>
  </div>
</DocumentFragment>
`;
